cookbook/case_studies/ml_training/mnist_classifier/pytorch_single_node_multi_gpu.py

```diff
@@ -0,0 +1,382 @@
+# `in pytorch docs <https://pytorch.org/tutorials/beginner/saving_loading_models.html#saving-and-loading-models>`_.
+# A common convention is to have the ``.pt`` extension for the model file.
+#
+# .. note::
+#    Note the usage of ``requests=Resources(gpu=WORLD_SIZE)``. This will force Flyte to allocate this task onto a
+#    machine with GPU(s), which in our case is 2 gpus. The task will be queued up until a machine with GPU(s) can be
+#    procured. Also, for the GPU Training to work, the Dockerfile needs to be built as explained in the
+#    :ref:`pytorch-dockerfile` section.
+
+# %%
+# Defining the ``task``
+# =====================
+#
+# Next we define the flyte task that kicks off the distributed training process. Here we call the
+# pytorch :ref:`multiprocessing <pytorch:torch.multiprocessing.spawn>` function to initiate a process on each
+# available GPU. Since we're parallelizing the data, each process will contain a copy of the model and pytorch
+# will handle syncing the weights across all processes on ``optimizer.step()`` calls.
+# 
+# See `here <https://pytorch.org/tutorials/beginner/dist_overview.html>`_ to read more about pytorch distributed
+# training.
+
+@task(
+    retries=2,
+    cache=True,
+    cache_version="1.2",
+    requests=Resources(gpu=str(WORLD_SIZE), mem="30Gi", storage="20Gi", ephemeral_storage="500Mi"),
+    limits=Resources(gpu=str(WORLD_SIZE), mem="30Gi", storage="20Gi", ephemeral_storage="500Mi"),
+)
+def pytorch_mnist_task(hp: Hyperparameters) -> TrainingOutputs:
+    print("Start MNIST training:")
+
+    world_size = torch.cuda.device_count()
+    print(f"Device count: {world_size}")
+    download_mnist(DATA_DIR)
+    mp.spawn(
+        train_mnist,
+        args=(world_size, hp),
+        nprocs=world_size,
+        join=True,
+    )
+    print("Training Complete")
+    with open(ACCURACIES_FILE) as fp:
+        accuracies = json.load(fp)
+    return TrainingOutputs(
+        epoch_accuracies=accuracies, model_state=PythonPickledFile(MODEL_FILE)
+    )
+
+
+
+# %%
+# Finally, we define a workflow to run the training algorithm. We return the model and accuracies.
+@workflow
+def pytorch_training_wf(hp: Hyperparameters = Hyperparameters(epochs=10, batch_size=128)) -> TrainingOutputs:
+    return pytorch_mnist_task(hp=hp)
+
+
+# %%
+# Running the Model Locally
+# =========================
+#
+# It is possible to run the model locally with almost no modifications (as long as the code takes care of resolving
+# if the code is distributed or not). This is how we can do it:
+if __name__ == "__main__":
+    model, accuracies = pytorch_training_wf(hp=Hyperparameters(epochs=10, batch_size=128))
+    print(f"Model: {model}, Accuracies: {accuracies}")
+
+# %%
+# Weights & Biases Report
+# =======================
+#
+# You can refer to the complete ``wandb`` report `here <https://wandb.ai/niels-bantilan/mnist-single-node-multi-gpu/reports/Pytorch-Single-node-Multi-GPU-Report--Vmlldzo5Mjk4Nzk>`__.
```

Comment:
Can we also have the accuracy, test_loss,... graphs in this report? We can also have the wandb table.

Reply:
sure, I can update that, tho it shouldn't block merging this PR